db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/PredicateList.java

```diff
@@ -564,6 +612,58 @@ public static Integer isIndexUseful(Predicate pred,
         return indexPosition;
     }
 
+    // Return an empty list if relOp doesn't match index expression. This is OK because an index expression must
+    // have at least one column reference. Also, since an index is defined on one table, an index expression
+    // cannot have column references referencing different tables.
+    private static List<ColumnReference> matchIndexExpression(RelationalOperator relOp, InListOperatorNode inNode,
+                                                              boolean isIn, boolean isProbe,
+                                                              ValueNode indexExprAst, Optimizable optTable)
+            throws StandardException
+    {
+        CollectNodesVisitor cnv = new CollectNodesVisitor(ColumnReference.class);
+        int tableNumber = optTable.getTableNumber();  // OK to be -1, will be checked when use
+        if (isIn) {
+            if (inNode.getLeftOperand().equals(indexExprAst)) {
+                inNode.getLeftOperand().accept(cnv);
+                if (isProbe) {
+                    assert relOp instanceof BinaryOperatorNode;
+                    BinaryOperatorNode binOp = (BinaryOperatorNode) relOp;
+                    binOp.setMatchIndexExpr(tableNumber, true);
+                }
+            }
+        } else {
+            if (relOp instanceof BinaryOperatorNode) {
+                BinaryOperatorNode binOp = (BinaryOperatorNode) relOp;
+                if (binOp.getLeftOperand().equals(indexExprAst)) {
+                    binOp.getLeftOperand().accept(cnv);
+                    binOp.setMatchIndexExpr(tableNumber, true);
+                } else if (binOp.getRightOperand().equals(indexExprAst)) {
+                    binOp.getRightOperand().accept(cnv);
+                    binOp.setMatchIndexExpr(tableNumber, false);
+                }
+            } else if (relOp instanceof IsNullNode) {
+                IsNullNode isNull = (IsNullNode) relOp;
+                if (isNull.getOperand().equals(indexExprAst)) {
+                    isNull.getOperand().accept(cnv);
+                    // No need to set any matchIndexExpr flag since it won't be used in code generation.
+                }
+            }
+        }
+        return cnv.getList();
```

Comment:
Right, the list is not useful. It can be optimized a bit. I made it that way just to reuse the call site logic with minimal change. In next commit, I turn the return value to boolean and tuned the caller a bit to handle that.